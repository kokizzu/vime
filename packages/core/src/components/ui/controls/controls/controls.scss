vime-controls {
  /**
   * @prop --controls-padding: The padding inside the controls.
   * @prop --controls-bg: The background color of the controls.
   * @prop --controls-spacing: The space between controls.
   */

  display: flex;
  position: absolute;
  flex-wrap: wrap;
  pointer-events: auto;
  z-index: $z-index-controls;
  background: var(--controls-bg);
  padding: var(--controls-padding);
  @include invisible();

  &.audio {
    position: relative;
  }
  
  &.hidden {
    display: none;
  }

  &.active {
    @include visible();
  }

  &.fullWidth {
    width: 100%;
  }
  
  &.fullHeight {
    height: 100%;
  }

  & > *:not(vime-control-group) {
    margin-left: var(--controls-spacing);

    &:first-child {
      margin-left: 0;
    }
  }
}