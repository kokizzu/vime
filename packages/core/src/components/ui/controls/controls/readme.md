# vime-controls

Responsible for positioning and laying out individual/groups of controls.

## Example

```html
<vime-player>
  <!-- ... -->
  <vime-ui>
    <vime-controls pin="bottomLeft" active-duration="3000" full-width>
      <!-- ... -->
    </vime-controls>
  </vime-ui>
</vime-player>
```

<!-- Auto Generated Below -->


## Properties

| Property               | Attribute                 | Description                                                                                                                 | Type                                                                                  | Default        |
| ---------------------- | ------------------------- | --------------------------------------------------------------------------------------------------------------------------- | ------------------------------------------------------------------------------------- | -------------- |
| `activeDuration`       | `active-duration`         | The length in milliseconds that the controls are active for before fading out. Audio players are not effected by this prop. | `number`                                                                              | `2750`         |
| `align`                | `align`                   | Sets the `align-items` flex property that aligns the individual controls on the cross-axis.                                 | `"center" \| "end" \| "start"`                                                        | `'center'`     |
| `direction`            | `direction`               | Sets the `flex-direction` property that manages the direction in which the controls are layed out.                          | `"column" \| "row"`                                                                   | `'row'`        |
| `fullHeight`           | `full-height`             | Whether the controls container should be 100% height. This has no effect if the view is of type `audio`.                    | `boolean`                                                                             | `false`        |
| `fullWidth`            | `full-width`              | Whether the controls container should be 100% width. This has no effect if the view is of type `audio`.                     | `boolean`                                                                             | `false`        |
| `hidden`               | `hidden`                  | Whether the controls are visible or not.                                                                                    | `boolean`                                                                             | `false`        |
| `hideOnMouseLeave`     | `hide-on-mouse-leave`     | Whether the controls should hide when the mouse leaves the player. Audio players are not effected by this prop.             | `boolean`                                                                             | `false`        |
| `hideWhenPaused`       | `hide-when-paused`        | Whether the controls should show/hide when paused. Audio players are not effected by this prop.                             | `boolean`                                                                             | `false`        |
| `justify`              | `justify`                 | Sets the `justify-content` flex property that aligns the individual controls on the main-axis.                              | `"center" \| "end" \| "space-around" \| "space-between" \| "space-evenly" \| "start"` | `'start'`      |
| `pin`                  | `pin`                     | Pins the controls to the defined position inside the video player. This has no effect when the view is of type `audio`.     | `"bottomLeft" \| "bottomRight" \| "center" \| "topLeft" \| "topRight"`                | `'bottomLeft'` |
| `waitForPlaybackStart` | `wait-for-playback-start` | Whether the controls should wait for playback to start before being shown. Audio players are not effected by this prop.     | `boolean`                                                                             | `false`        |


## Slots

| Slot | Description               |
| ---- | ------------------------- |
|      | Used to pass in controls. |


## CSS Custom Properties

| Name                 | Description                           |
| -------------------- | ------------------------------------- |
| `--controls-bg`      | The background color of the controls. |
| `--controls-padding` | The padding inside the controls.      |
| `--controls-spacing` | The space between controls.           |


## Dependencies

### Used by

 - [vime-default-controls](../default-controls)

### Graph
```mermaid
graph TD;
  vime-default-controls --> vime-controls
  style vime-controls fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
