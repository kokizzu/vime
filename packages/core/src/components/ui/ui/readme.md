# vime-ui

Simple container that holds a collection of user interface components.

The only important role this component really has is, avoiding overlapping custom UI with the 
native iOS media player UI. Therefore, custom UI is only displayed on iOS if the `playsinline` prop 
is `true`, and the player is not in fullscreen mode.

## Example

```html
<vime-player>
  <!-- ... -->

  <vime-ui>
    <vime-poster />
    <!-- ... -->
  </vime-ui>
</vime-player>
```

<!-- Auto Generated Below -->


----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
