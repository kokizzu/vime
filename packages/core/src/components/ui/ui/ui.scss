vime-ui {
  position: relative;
  pointer-events: none;
  z-index: $z-index-ui;
  width: 100%;

  &.hidden {
    display: none;
  }

  &.video {
    position: absolute;
    top: 0;
    left: 0;
    height: 100%;
  }
}