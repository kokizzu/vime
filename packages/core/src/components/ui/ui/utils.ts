import { getElement } from '@stencil/core';
import { HTMLStencilElement } from '@stencil/core/internal';

export const findUIRoot = (ref: any) => {
  let ui = getElement(ref);

  while (!(/^VIME-UI$/.test(ui?.nodeName))) {
    ui = ui.parentElement as HTMLStencilElement;
  }

  return ui as HTMLVimeUiElement;
};
