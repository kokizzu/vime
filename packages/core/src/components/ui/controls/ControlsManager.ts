import { getElement } from '@stencil/core';
import { HTMLStencilElement } from '@stencil/core/internal';
import { ViewType } from '../../core/player/ViewType';
import { createPlayerStateDispatcher, PlayerStateDispatcher } from '../../core/player/PlayerState';
import { listen } from '../../../utils/dom';
import { Disposal } from '../../core/player/Disposal';
import { PlayerProp } from '../../core/player/PlayerProp';

export class ControlsManager {
  private dispatch?: PlayerStateDispatcher;

  private hideControlsTimeout?: number;

  private disposal = new Disposal();

  private isEnabled = false;

  private isPaused = true;

  private hasPlaybackStarted = false;

  private view?: ViewType;

  set enabled(enabled: boolean) {
    this.isEnabled = enabled;
    this.onControlsChange();
  }

  set paused(paused: boolean) {
    this.isPaused = paused;
    this.onControlsChange();
  }

  set viewType(viewType: ViewType | undefined) {
    this.view = viewType;
    this.onControlsChange();
  }

  set playbackStarted(started: boolean) {
    this.hasPlaybackStarted = started;
    this.onControlsChange();
  }

  constructor(
    ref?: any,
    private readonly duration = 2500,
    private readonly showsControlsBeforePlaybackStart = true,
    private readonly lockControlsWhenPaused = true,
    private readonly hideControlsOnMouseLeave = true,
  ) {
    const el = getElement(ref);
    this.dispatch = createPlayerStateDispatcher(ref);

    let player = el;
    while (player?.nodeName !== 'VIME-PLAYER') {
      player = player.parentNode as HTMLStencilElement;
    }

    const events = ['focus', 'keydown', 'click', 'mousemove', 'touchstart', 'mouseleave'];
    events.forEach((event) => {
      this.disposal.add(listen(player, event, this.onControlsChange.bind(this)));
    });
  }

  private show() {
    this.dispatch!(PlayerProp.IsControlsActive, true);
  }

  private hide() {
    this.dispatch!(PlayerProp.IsControlsActive, false);
  }

  private hideWithDelay() {
    window.requestAnimationFrame(() => {
      this.hideControlsTimeout = window.setTimeout(() => {
        this.hide();
      }, this.duration);
    });
  }

  onControlsChange(e?: Event) {
    window.clearTimeout(this.hideControlsTimeout);

    if (!this.isEnabled) {
      this.hide();
      return;
    }

    if (this.view === ViewType.Audio) {
      this.show();
      return;
    }

    if (this.view !== ViewType.Video) {
      this.hide();
      return;
    }

    if (this.showsControlsBeforePlaybackStart && !this.hasPlaybackStarted) {
      this.show();
      return;
    }

    if (!this.showsControlsBeforePlaybackStart && !this.hasPlaybackStarted) {
      this.hide();
      return;
    }

    if (
      (e?.type === 'mouseleave')
      && this.hideControlsOnMouseLeave
      && !(this.lockControlsWhenPaused && this.isPaused)
    ) {
      this.hide();
      return;
    }

    if (!this.isPaused || !this.lockControlsWhenPaused) {
      this.show();
      this.hideWithDelay();
      return;
    }

    (this.isPaused) ? this.show() : this.hide();
  }

  destroy() {
    this.disposal.empty();
    this.dispatch = undefined;
  }
}
