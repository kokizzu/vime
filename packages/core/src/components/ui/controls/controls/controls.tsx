import {
  h, Host, Component,
  Prop, Element, Watch,
  State,
} from '@stencil/core';
import { openPlayerWormhole } from '../../../core/player/PlayerWormhole';
import { PlayerProp, PlayerProps } from '../../../core/player/PlayerProp';
import { PlayerStateDispatcher, createPlayerStateDispatcher } from '../../../core/player/PlayerState';
import { Disposal } from '../../../core/player/Disposal';
import { listen, isColliding } from '../../../../utils/dom';
import { isNull } from '../../../../utils/unit';
import { debounce } from '../../../../utils/timing';
import { findMyPlayer } from '../../../core/player/utils';

/**
 * We want to keep the controls active state in-sync per player.
 */
const playerRef: Record<any, HTMLVimePlayerElement> = {};
const hideControlsTimeout: Record<any, number | undefined> = {};

/**
 * @slot - Used to pass in controls.
 */
@Component({
  tag: 'vime-controls',
  styleUrl: 'controls.scss',
})
export class Controls {
  private dispatch?: PlayerStateDispatcher;

  private disposal = new Disposal();

  @Element() el!: HTMLVimeControlsElement;

  @State() isInteracting = false;

  /**
   * Whether the controls are visible or not.
   */
  @Prop() hidden = false;

  /**
   * Whether the controls container should be 100% width. This has no effect if the view is of
   * type `audio`.
   */
  @Prop() fullWidth = false;

  /**
   * Whether the controls container should be 100% height. This has no effect if the view is of
   * type `audio`.
   */
  @Prop() fullHeight = false;

  /**
   * Sets the `flex-direction` property that manages the direction in which the controls are layed
   * out.
   */
  @Prop() direction: 'row' | 'column' = 'row';

  /**
   * Sets the `align-items` flex property that aligns the individual controls on the cross-axis.
   */
  @Prop() align: 'start' | 'center' | 'end' = 'center';

  /**
   * Sets the `justify-content` flex property that aligns the individual controls on the main-axis.
   */
  @Prop() justify: 'start'
  | 'center'
  | 'end'
  | 'space-around'
  | 'space-between'
  | 'space-evenly' = 'start';

  /**
   * Pins the controls to the defined position inside the video player. This has no effect when
   * the view is of type `audio`.
   */
  @Prop() pin: 'topLeft' | 'topRight' | 'bottomLeft' | 'bottomRight' | 'center' = 'bottomLeft';

  /**
   * The length in milliseconds that the controls are active for before fading out. Audio players
   * are not effected by this prop.
   */
  @Prop() activeDuration = 2750;

  /**
   * Whether the controls should wait for playback to start before being shown. Audio players
   * are not effected by this prop.
   */
  @Prop() waitForPlaybackStart = false;

  /**
   * Whether the controls should show/hide when paused. Audio players are not effected by this prop.
   */
  @Prop() hideWhenPaused = false;

  /**
   * Whether the controls should hide when the mouse leaves the player. Audio players are not
   * effected by this prop.
   */
  @Prop() hideOnMouseLeave = false;

  /**
   * @internal
   */
  @Prop() isAudioView: PlayerProps[PlayerProp.IsAudioView] = false;

  /**
   * @internal
   */
  @Prop() playbackReady: PlayerProps[PlayerProp.PlaybackReady] = false;

  /**
   * @internal
   */
  @Prop() isControlsActive: PlayerProps[PlayerProp.IsControlsActive] = false;

  /**
   * @internal
   */
  @Prop() paused: PlayerProps[PlayerProp.Paused] = true;

  /**
   * @internal
   */
  @Prop() playbackStarted: PlayerProps[PlayerProp.PlaybackStarted] = false;

  connectedCallback() {
    this.dispatch = createPlayerStateDispatcher(this);
    this.setupPlayerListeners();
    this.checkForCaptionsCollision();
  }

  componentDidRender() {
    this.checkForCaptionsCollision();
  }

  disconnectedCallback() {
    this.disposal.empty();
    delete hideControlsTimeout[playerRef[this]];
    delete playerRef[this];
  }

  private setupPlayerListeners() {
    const player = findMyPlayer(this);
    const events = ['focus', 'keydown', 'click', 'mousemove', 'touchstart', 'mouseleave'];
    const callback = debounce(this.onControlsChange, 50, true).bind(this);
    events.forEach((event) => { this.disposal.add(listen(player, event, callback)); });
    // @ts-ignore
    playerRef[this] = player;
  }

  private checkForCaptionsCollision() {
    const captions = this.el.parentNode!.querySelector('vime-captions') as HTMLVimeCaptionsElement;
    if (isNull(captions)) return;
    captions!.controlsHeight = isColliding(this.el, captions)
      ? parseFloat(window.getComputedStyle(this.el).height)
      : 0;
  }

  private show() {
    this.dispatch!(PlayerProp.IsControlsActive, true);
  }

  private hide() {
    this.dispatch!(PlayerProp.IsControlsActive, false);
  }

  private hideWithDelay() {
    hideControlsTimeout[playerRef[this]] = setTimeout(() => {
      this.hide();
    }, this.activeDuration) as any;
  }

  @Watch('paused')
  @Watch('hidden')
  @Watch('isAudioView')
  @Watch('isInteracting')
  @Watch('hideWhenPaused')
  @Watch('hideOnMouseLeave')
  @Watch('playbackStarted')
  @Watch('waitForPlaybackStart')
  @Watch('playbackReady')
  private onControlsChange(event?: Event) {
    // @ts-ignore
    clearTimeout(hideControlsTimeout[playerRef[this]]);

    if (this.hidden || !this.playbackReady) {
      this.hide();
      return;
    }

    if (this.isAudioView) {
      this.show();
      return;
    }

    if (this.waitForPlaybackStart && !this.playbackStarted) {
      this.hide();
      return;
    }

    if (this.isInteracting) {
      this.show();
      return;
    }

    if (this.hideWhenPaused && this.paused) {
      this.hideWithDelay();
      return;
    }

    if (this.hideOnMouseLeave && (event?.type === 'mouseleave')) {
      this.hide();
      return;
    }

    if (!this.paused) {
      this.show();
      this.hideWithDelay();
      return;
    }

    this.show();
  }

  private getPosition() {
    if (this.isAudioView) return {};

    if (this.pin === 'center') {
      return {
        top: '50%',
        left: '50%',
        transform: 'translate(-50%, -50%)',
      };
    }

    // topLeft => { top: 0, left: 0 }
    const pos = this.pin.split(/(?=[L|R])/).map((s) => s.toLowerCase());
    return { [pos[0]]: 0, [pos[1]]: 0 };
  }

  private onStartInteraction() {
    this.isInteracting = true;
  }

  private onEndInteraction() {
    this.isInteracting = false;
  }

  render() {
    return (
      <Host
        style={{
          ...this.getPosition(),
          flexDirection: this.direction,
          alignItems: (this.align === 'center') ? 'center' : `flex-${this.align}`,
          justifyContent: this.justify,
        }}
        class={{
          audio: this.isAudioView,
          hidden: this.hidden,
          active: this.playbackReady && this.isControlsActive,
          fullWidth: this.isAudioView || this.fullWidth,
          fullHeight: !this.isAudioView && this.fullHeight,
        }}
        onMouseEnter={this.onStartInteraction.bind(this)}
        onMouseLeave={this.onEndInteraction.bind(this)}
        onTouchStart={this.onStartInteraction.bind(this)}
        onTouchEnd={this.onEndInteraction.bind(this)}
      >
        <slot />
      </Host>
    );
  }
}

openPlayerWormhole(Controls, [
  PlayerProp.IsAudioView,
  PlayerProp.PlaybackReady,
  PlayerProp.IsControlsActive,
  PlayerProp.Paused,
  PlayerProp.PlaybackStarted,
]);
