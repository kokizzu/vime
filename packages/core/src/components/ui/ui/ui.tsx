import {
  h, Host, Component, Prop,
} from '@stencil/core';
import { openPlayerWormhole } from '../../core/player/PlayerWormhole';
import { PlayerProp, PlayerProps } from '../../core/player/PlayerProp';
import { IS_IOS } from '../../../utils/support';

@Component({
  tag: 'vime-ui',
  styleUrl: 'ui.scss',
})
export class UI {
  /**
   * @internal
   */
  @Prop() isVideoView: PlayerProps[PlayerProp.IsVideoView] = false;

  /**
   * @internal
   */
  @Prop() playsinline: PlayerProps[PlayerProp.Playsinline] = false;

  /**
   * @internal
   */
  @Prop() isFullscreenActive: PlayerProps[PlayerProp.IsFullscreenActive] = false;

  render() {
    const canShowCustomUI = !IS_IOS
      || !this.isVideoView
      || (this.playsinline && !this.isFullscreenActive);

    return (
      <Host
        class={{
          hidden: !canShowCustomUI,
          video: this.isVideoView,
        }}
      >
        {canShowCustomUI && <slot />}
      </Host>
    );
  }
}

openPlayerWormhole(UI, [
  PlayerProp.IsVideoView,
  PlayerProp.Playsinline,
  PlayerProp.IsFullscreenActive,
]);
