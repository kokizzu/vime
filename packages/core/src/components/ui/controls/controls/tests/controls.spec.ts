import { SpecPage } from '@stencil/core/testing';
import { Controls } from '../controls';
import { newUISpecPage } from '../../../ui/tests';
import { PlayerProp } from '../../../../core/player/PlayerProp';
import { ViewType } from '../../../../core/player/ViewType';

let page: SpecPage;
let provider: HTMLVimeFaketubeElement;
let controls: HTMLVimeControlsElement;

jest.useFakeTimers();

beforeEach(async () => {
  ({ page, provider } = await newUISpecPage(
    [Controls],
    '<vime-controls />',
  ));

  controls = page.root!.querySelector('vime-controls')!;
});

it('should be structurally sound', () => {
  expect(controls).toMatchSnapshot();
});

it('should hide controls', async () => {
  expect(controls).not.toHaveClass('hidden');
  controls.hidden = true;
  await page.waitForChanges();
  expect(controls).toHaveClass('hidden');
});

it('should become active', async () => {
  expect(controls).not.toHaveClass('active');
  await provider.dispatchStateChange(PlayerProp.viewType, ViewType.Video);
  await provider.dispatchStateChange(PlayerProp.playbackReady, true);
  await provider.dispatchStateChange(PlayerProp.isControlsActive, true);
  await page.waitForChanges();
  await page.waitForChanges();
  expect(controls).toHaveClass('active');
});

it('should only be active when audio view', async () => {
  expect(controls).not.toHaveClass('active');
  await provider.dispatchStateChange(PlayerProp.viewType, ViewType.Audio);
  await provider.dispatchStateChange(PlayerProp.playbackReady, true);
  await page.waitForChanges();
  await page.waitForChanges();
  expect(controls).toHaveClass('active');
  await provider.dispatchStateChange(PlayerProp.paused, false);
  await page.waitForChanges();
  await page.waitForChanges();
  jest.runAllTimers();
  expect(controls).toHaveClass('active');
});

it('should apply full width', async () => {
  expect(controls).not.toHaveClass('fullWidth');
  controls.fullWidth = true;
  await page.waitForChanges();
  expect(controls).toHaveClass('fullWidth');
});

it('should apply full height', async () => {
  expect(controls).not.toHaveClass('fullHeight');
  controls.fullHeight = true;
  await page.waitForChanges();
  expect(controls).toHaveClass('fullHeight');
});

it('should apply direction', async () => {
  expect(controls.style.flexDirection).toEqual('row');
  controls.direction = 'column';
  await page.waitForChanges();
  expect(controls.style.flexDirection).toEqual('column');
});

it('should align controls', async () => {
  expect(controls.style.alignItems).toEqual('center');
  controls.align = 'start';
  await page.waitForChanges();
  expect(controls.style.alignItems).toEqual('flex-start');
  controls.align = 'end';
  await page.waitForChanges();
  expect(controls.style.alignItems).toEqual('flex-end');
});

it('should justify controls', async () => {
  expect(controls.style.justifyContent).toEqual('start');
  controls.justify = 'space-between';
  await page.waitForChanges();
  expect(controls.style.justifyContent).toEqual('space-between');
});

it('should not pin controls if audio', async () => {
  await provider.dispatchStateChange(PlayerProp.viewType, ViewType.Audio);
  controls.pin = 'topLeft';
  await page.waitForChanges();
  expect(controls.style.top).toEqual('');
  expect(controls.style.left).toEqual('');
});

it('should pin controls to the center', async () => {
  await provider.dispatchStateChange(PlayerProp.viewType, ViewType.Video);
  controls.pin = 'center';
  await page.waitForChanges();
  expect(controls.style.top).toEqual('50%');
  expect(controls.style.left).toEqual('50%');
});

it('should pin controls', async () => {
  await provider.dispatchStateChange(PlayerProp.viewType, ViewType.Video);
  controls.pin = 'topLeft';
  await page.waitForChanges();
  expect(controls.style.top).toEqual('0');
  expect(controls.style.left).toEqual('0');
  controls.pin = 'bottomRight';
  await page.waitForChanges();
  expect(controls.style.top).toEqual('');
  expect(controls.style.left).toEqual('');
  expect(controls.style.bottom).toEqual('0');
  expect(controls.style.right).toEqual('0');
});

it('should change active duration', async () => {
  controls.activeDuration = 3500;
  await provider.dispatchStateChange(PlayerProp.playbackReady, true);
  await provider.dispatchStateChange(PlayerProp.viewType, ViewType.Video);
  await provider.dispatchStateChange(PlayerProp.isControlsActive, true);
  await provider.dispatchStateChange(PlayerProp.paused, false);
  await page.waitForChanges();
  await page.waitForChanges();
  expect(controls.isControlsActive).toBeTruthy();
  requestAnimationFrame(async () => {
    jest.advanceTimersByTime(3000);
    await page.waitForChanges();
    expect(controls.isControlsActive).toBeTruthy();
    jest.advanceTimersByTime(500);
    await page.waitForChanges();
    expect(controls.isControlsActive).toBeFalsy();
  });
});

it('should wait for playback to start before showing controls', async () => {
  controls.waitForPlaybackStart = true;
  await provider.dispatchStateChange(PlayerProp.playbackReady, true);
  await provider.dispatchStateChange(PlayerProp.viewType, ViewType.Video);
  await provider.dispatchStateChange(PlayerProp.paused, false);
  await page.waitForChanges();
  await page.waitForChanges();
  expect(controls.isControlsActive).toBeFalsy();
  await provider.dispatchStateChange(PlayerProp.playbackStarted, true);
  await page.waitForChanges();
  await page.waitForChanges();
  expect(controls.isControlsActive).toBeTruthy();
});

it('should hide controls when paused', async () => {
  controls.hideWhenPaused = true;
  await provider.dispatchStateChange(PlayerProp.playbackReady, true);
  await provider.dispatchStateChange(PlayerProp.viewType, ViewType.Video);
  await provider.dispatchStateChange(PlayerProp.paused, false);
  await page.waitForChanges();
  await page.waitForChanges();
  expect(controls.isControlsActive).toBeTruthy();
  await provider.dispatchStateChange(PlayerProp.paused, true);
  await page.waitForChanges();
  await page.waitForChanges();
  requestAnimationFrame(async () => {
    expect(controls.isControlsActive).toBeTruthy();
    jest.runAllTimers();
    await page.waitForChanges();
    expect(controls.isControlsActive).toBeFalsy();
  });
});

it('should hide controls on mouse leave', async () => {
  controls.hideOnMouseLeave = true;
  await provider.dispatchStateChange(PlayerProp.playbackReady, true);
  await provider.dispatchStateChange(PlayerProp.viewType, ViewType.Video);
  await page.waitForChanges();
  await page.waitForChanges();
  expect(controls.isControlsActive).toBeTruthy();
  controls.dispatchEvent(new Event('mouseleave', { bubbles: true }));
  await page.waitForChanges();
  await page.waitForChanges();
  expect(controls.isControlsActive).toBeFalsy();
});
