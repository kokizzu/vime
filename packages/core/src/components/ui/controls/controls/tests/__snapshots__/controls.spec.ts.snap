// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should be structurally sound 1`] = `
<vime-controls style="bottom: 0; left: 0; flex-direction: row; align-items: center; justify-content: start;">
  <!---->
</vime-controls>
`;
